Fix a crash that occurred because LowerESIMDVecArg pass was trying to

pratikashar opened the pull request:
add code to a function declaration.

This fixes compilation crash.

Signed-off-by: pratik ashar pratik.j.ashar@intel.com

kbobrovs reviewed (changes requested):
Please add a test.

DenisBakhvalov reviewed (changes requested):
@pratikashar , I suggest creating an LLVM IR test in `llvm/test/SYCLLowerIR` and use `opt -LowerESIMDVecArg` to test the patch.

pratikashar commented:
@dendibakh - the bug was a crash during compilation. this checkin fixes the crash. i've used same input that caused the crash.

DenisBakhvalov commented:
> @dendibakh - the bug was a crash during compilation. this checkin fixes the crash. i've used same input that caused the crash.

I believe we should create a minimal test case that exposes the bug that was fixed. C++ test goes through many other steps that are irrelevant to the issue. LLVM IR test would go straight to the bug.

Getting the IR for the test case should be easy. You can add `-mllvm -print-after-all -mllvm -print-module-scope` to the compilation line. Since it crashes on the LowerESIMDVecArg pass, the last printed IR would be right before that pass. Then you can do a little cleanup or use `llvm-extract` to reduce the IR.

kbobrovs commented:
@pratikashar, the suggestion from @DenisBakhvalov makes sense to me - this makes the test case much faster. This is also a practice in LLVM to write .ll tests for bugfixes in the llvm passes. Please consider this refactoring.

Current test would be most valuable, I think, if moved to llvm-test-suite and converted to end-to-end test (compilation + execution + verification). But this is of course not necessary for this PR - just a thought.

pratikashar commented:
Given that this is a simple nullptr dereference related failure, I'm not sure adding a test is critical. Any trivial test that exercises extern functions would test this nullptr problem that I fixed in this PR.

But keeping in mind majority suggestion, I've added a LIT test for this scenario and removed earlier cpp test.

I see that buildbot Jenkins task failed, but I dont see the log. Also, Idk how to run LIT test locally since I couldnt find opt.exe. Can you suggest the command line to try this LIT test locally?

DenisBakhvalov commented:
> Also, Idk how to run LIT test locally since I couldnt find opt.exe. Can you suggest the command line to try this LIT test locally?

You can run `ninja check-llvm` or `make check-llvm`. Running individual tests is: `llvm-lit llvm/test/SYCLLowerIR/esimd_subroutine_extern.ll`

pratikashar commented:
Thanks @DenisBakhvalov. I was able to reproduce the target related error and have fixed the LIT test. I've also reduced it significantly locally. However, for some reason I see a crash locally:
While deleting: <16 x float> addrspace(4)* %
Use still stuck around after Def is destroyed:  <badref> = bitcast <16 x float> addrspace(4)* %0 to %"class._ZTSN2cl4sycl5INTEL3gpu4simdIfLi16EEE.cl::sycl::INTEL::gpu::simd" addrspace(4)*
Assertion failed: materialized_use_empty() && "Uses remain when a value is destroyed!", file D:\pratik\github-esimd\llvm\llvm\lib\IR\Value.cpp, line 100

This doesnt show up if I run opt.exe from VS IDE. So I've uploaded the new LIT test to see if it is a local build issue for me.

pratikashar commented:
@DenisBakhvalov, @kbobrovs - please approve/merge changes if everything looks good.

kbobrovs commented on llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp:
```diff
@@ -181,8 +181,12 @@ Function *ESIMDLowerVecArgPass::rewriteFunc(Function &F) {
+  // insert bitcasts in new function only if its a definition
   for (auto &B : BitCasts) {
-    NF->begin()->getInstList().push_front(B);
+    if (!F.isDeclaration())
```
why not to move this check to the loop above and avoid creation and addition to `BitCasts` in case F is declaration? Am I missing something? Also if F is a declaration, do we need to clone, or just creating a new declaration is enough?

pratikashar replied:
i'm not sure whether CloneFunctionInto is needed for declaration as well, i assumed it is. since it uses VMap and given that VMap is populated in the loop above it, i chose to use this fix.

kbobrovs replied:
I suggest to check this and remove redundant code if cloning is not really needed.

pratikashar replied:
it is needed as it adds attribute on the function declaration (dso_local).

kbobrovs replied:
Attribute can be added w/o using cloning, right?

pratikashar replied:
i prefer using existing API as it works for both definitions/declarations. in case llvm adds more fields for functions, this API would ensure automatic compliance.

kbobrovs commented:
OK, here is what I found in CloneFunctionInto:
```cpp
  // Everything else beyond this point deals with function instructions,
  // so if we are dealing with a function declaration, we're done.
  if (OldFunc->isDeclaration())
    return;
```
So cloning declarations is fine. But still, I think you should not be creating objects then undoing it later. The check should be moved above. Bitcasts must not be created if F is a decl.

pratikashar commented:
here is part of implementation of CloneFunctionInfo():

   // Clone any argument attributes that are present in the VMap.
   for (const Argument &OldArg : OldFunc->args()) {
     if (Argument *NewArg = dyn_cast<Argument>(VMap[&OldArg])) {
       NewArgAttrs[NewArg->getArgNo()] =
           OldAttrs.getParamAttributes(OldArg.getArgNo());
     }
   }
  
Notice it uses dyn_cast<Argument>(VMap[&OldArg]). dyn_cast internally uses isa<> which doesnt expect a nullptr.

i made following change to verify this:
      auto BitCast = F.isDeclaration() ? nullptr :
          new BitCastInst(NF->getArg(I), Arg->getType());

so for declaration we push nullptr in to VMap. and i found this causes a crash, as i expected.

kbobrovs commented:
> here is part of implementation of CloneFunctionInfo():
> 
> // Clone any argument attributes that are present in the VMap.
> for (const Argument &OldArg : OldFunc->args()) {
> if (Argument *NewArg = dyn_cast(VMap[&OldArg])) {
> NewArgAttrs[NewArg->getArgNo()] =
> OldAttrs.getParamAttributes(OldArg.getArgNo());
> }
> }
> 
> Notice it uses dyn_cast(VMap[&OldArg]). dyn_cast internally uses isa<> which doesnt expect a nullptr.
> 
> i made following change to verify this:
> auto BitCast = F.isDeclaration() ? nullptr :
> new BitCastInst(NF->getArg(I), Arg->getType());
> 
> so for declaration we push nullptr in to VMap. and i found this causes a crash, as i expected.

I see, thanks for investigating. Seems like a drawback of LLVM cloning interface if it requires dummy mappings.

DenisBakhvalov commented on llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp:
```diff
@@ -181,8 +181,12 @@ Function *ESIMDLowerVecArgPass::rewriteFunc(Function &F) {
+    if (!F.isDeclaration())
+      NF->begin()->getInstList().push_front(B);
+    else
+      delete B;
```
I think you can put the `if (!F.isDeclaration())` check in the previous loop where you populate the `BitCasts` vector. This way you don't need to `delete B`.

pratikashar replied:
@dendibakh - please refer to discussion above discussing exactly this.

DenisBakhvalov replied:
@pratikashar - ah, OK.
